### PR TITLE
add query builder

### LINKED ISSUE #172: WS2024: Add an advanced search interface to Facebook
While Facebook contains flags and other tools to filter searches by state, town name, class year, etc., there is no interface to do so beyond the search bar itself. Most search tools will have an "Advanced Search" interface to make these searches user friendly, but Facebook lacks such a feature. 

I would like to add a button to the WSO website under "Search" called "Advanced Search". When clicked, this button will open a page, with text inputs corresponding to filtering searches by keyword, student dorm, country, town name, class year, etc. 

While this feature will not strictly add any new functionality, it is my hope that it will make the website more user-friendly.

### PR DESCRIPTION
**Pull request by asashepard**
Resolves #172

### CODE REVIEW

File: src/components/Homepage.tsx [hunk 1]
```diff
@@ -11,13 +12,182 @@ import { PostType } from "../lib/types";
...
-    navigateTo(`/facebook?q=${query}`);
+    if (selected) navigateTo(`/facebook?q=${query}`);
+    else navigateTo(`/facebook?q=${search}`);
   };
+  const [name, updateName] = useState("");
+  const [unix, updateUnix] = useState("");
+  const [country, updateCountry] = useState("");
+  const [state, updateState] = useState("");
+  const [city, updateCity] = useState("");
+  const [button1, setButton1] = useState(false);
```
Comment (AndrewMuh) [inline src/components/Homepage.tsx:30]: what is button1-7? Fix naming to like professorFilterSelected, right now this makes the code very difficult to understand

File: src/components/Homepage.tsx [hunk 2]
```diff
@@ -11,13 +12,182 @@ import { PostType } from "../lib/types";
...
+  const [button2, setButton2] = useState(false);
+  const [button3, setButton3] = useState(false);
+  const [button4, setButton4] = useState(false);
+  const [button5, setButton5] = useState(false);
+  const [button6, setButton6] = useState(false);
+  const [button7, setButton7] = useState(false);
+  const [building, updateBuilding] = useState("");
+
+  function handleClick() {
```
Comment (AndrewMuh) [inline src/components/Homepage.tsx:39]: Change to `const handleClick = () => {}` instead of using function keyword. Do this for all functions.

File: src/components/Homepage.tsx [hunk 3]
```diff
@@ -11,13 +12,182 @@ import { PostType } from "../lib/types";
 
 const Homepage = () => {
   const navigateTo = useNavigate();
+  const [search, updateSearch] = useState("");
   const [query, updateQuery] = useState("");
+  const [selected, setSelected] = useState(false);
```
Comment (AndrewMuh) [inline src/components/Homepage.tsx:17]: rename selected to something that more closely identifies what's selected

File: src/components/Homepage.tsx [hunk 4]
```diff
@@ -11,13 +12,182 @@ import { PostType } from "../lib/types";
...
+
+    // base query (search box)
+    if (search.trim() !== "") {
+      str += search;
+      prior = true;
+    }
+
+    // name
+    if (name.trim() !== "") {
+      str += prior ? " AND (" : "(";
```
Comment (AndrewMuh) [inline src/components/Homepage.tsx:67]: You can use format string for cleaner building.

File: src/components/Homepage.tsx [hunk 5]
```diff
@@ -11,13 +12,182 @@ import { PostType } from "../lib/types";
...
+    let prior = false;
+
+    // base query (search box)
+    if (search.trim() !== "") {
+      str += search;
+      prior = true;
+    }
+
+    // name
+    if (name.trim() !== "") {
```
Comment (AndrewMuh) [inline src/components/Homepage.tsx:66]: break each one of these out to it's own function. This function is too long

File: src/components/Homepage.tsx [hunk 6]
```diff
@@ -11,13 +12,182 @@ import { PostType } from "../lib/types";
...
+      <button
+        onClick={handleClick}
+        className={selected ? "button-toggled" : "button-default"}
+      >
+        Advanced
+      </button>
+    );
+  }
+
+  function QueryManager() {
```
Comment (AndrewMuh) [inline src/components/Homepage.tsx:54]: I would actually make query manager its own component and just use it in the homepage.

File: src/components/Homepage.tsx [hunk 7]
```diff
@@ -38,20 +208,234 @@ const Homepage = () => {
...
+                  </td>
+                  <td align="right">
+                    <strong>state</strong>
+                  </td>
+                  <td align="center">
+                    <div className="vertical">
+                      <select
+                        onChange={(event) => updateState(event.target.value)}
+                      >
+                        <option value=""></option>
```
Comment (AndrewMuh) [inline src/components/Homepage.tsx:276]: This should break out to it's own helper function.

File: src/components/views/Facebook/FacebookLayout.tsx
Comment (AndrewMuh) [inline src/components/views/Facebook/FacebookLayout.tsx:1]: Why is the code form homepage all here? If you find yourself duplicating code like this, you should definitely be breaking stuff into its own components.

**Review by AndrewMuh — CHANGES_REQUESTED** [review]
Make code more readable and break things into smaller components/functions. Rerequest me when done.

**Comment by yechs** [comment]
Resolves #172

File: src/components/Homepage.tsx [hunk 8]
```diff
@@ -11,12 +13,45 @@ import { PostType } from "../lib/types";
...
+  };
+  const [name, updateName] = useState("");
+  const [unix, updateUnix] = useState("");
+  const [country, updateCountry] = useState("");
+  const [state, updateState] = useState("");
+  const [city, updateCity] = useState("");
+  const [year27Selected, setYear27Selected] = useState(false);
+  const [year26Selected, setYear26Selected] = useState(false);
+  const [year25Selected, setYear25Selected] = useState(false);
+  const [year24Selected, setYear24Selected] = useState(false);
```
Comment (yechs) [inline src/components/Homepage.tsx:34]: The years can be autogenerated. These states should be interpreted as first-years/sophomores/juniors/seniors, and then added by the base academic year.

The academic year can be calculated with these codes here. We may want to turn it into a util function for reuse.

https://github.com/WilliamsStudentsOnline/wso-react/blob/d82b8cee4f4442373459dbdc01f8d4ced20e845c/src/constants/constants.js#L25-L31

File: src/components/Homepage.tsx [hunk 9]
```diff
@@ -11,12 +13,45 @@ import { PostType } from "../lib/types";
 
 const Homepage = () => {
   const navigateTo = useNavigate();
+  const [search, updateSearch] = useState("");
   const [query, updateQuery] = useState("");
```
Comment (yechs) [inline src/components/Homepage.tsx:17]: This is an interesting design. However, I think `query` and `search` can be merged as one, where updates directly shows up in the query box.

That being said, we can offer user a separate textbox in the advanced search options.

Think of the "all these words: " box in [Google's Advanced Search page](https://www.google.com/advanced_search).

File: src/components/QueryManager.jsx [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+  if (year27Selected || year26Selected || year25Selected || year24Selected)
+    str += prior ? " AND (" : "(";
+  if (year27Selected) str += 'year: "27"';
+  if (year26Selected) str += firstYear === 2 ? 'year: "26"' : ' OR year: "26"';
+  if (year25Selected) str += firstYear === 3 ? 'year: "25"' : ' OR year: "25"';
+  if (year24Selected) str += firstYear === 4 ? 'year: "24"' : ' OR year: "24"';
+  if (year27Selected || year26Selected || year25Selected || year24Selected) {
+    str += ")";
+    prior = true;
+  }
```
Comment (yechs) [inline src/components/QueryManager.jsx:90]: As in previous comment, turn the state into a mathematical term to be added to the base academic year.

File: src/components/Homepage.tsx (hunk 8 above)
Comment (yechs) [reply src/components/Homepage.tsx:34]: Also, the state here could simply be a number, or an enum of four options (preferred) if you want.

File: src/components/views/Facebook/FacebookLayout.tsx [hunk 11]
```diff
@@ -12,7 +14,24 @@ const FacebookLayout = ({ children }: { children: React.ReactElement }) => {
...
+  const [state, updateState] = useState("");
+  const [city, updateCity] = useState("");
+  const [year27Selected, setYear27Selected] = useState(false);
+  const [year26Selected, setYear26Selected] = useState(false);
+  const [year25Selected, setYear25Selected] = useState(false);
+  const [year24Selected, setYear24Selected] = useState(false);
+  const [professorSelected, setProfessorSelected] = useState(false);
+  const [staffSelected, setStaffSelected] = useState(false);
+  const [studentSelected, setStudentSelected] = useState(false);
+  const [building, updateBuilding] = useState("");
```
Comment (yechs) [inline src/components/views/Facebook/FacebookLayout.tsx:34]: Having these states stored duplicated at two places doesn't seem ideal.

We can maybe use a global state (like a redux slice that isn't persisted), or manually pass these down as props (would be a lot of painful work IMO)

 @AndrewMuh thoughts?

**Review by yechs — CHANGES_REQUESTED** [review]
Some of the feedback was given synchronously during the dev session.

TL;DR
- [ ] state management without duplication
- [ ] automatic class year calculation (no hardcoded class years)

**Comment by charlietharas** [comment]
TODO:
- update branch
- visual improvements
